### PR TITLE
fix: oracle price decimals [0f-ARK-011]

### PR DESCRIPTION
**Pull request by nieldeckx**

### CODE REVIEW

**Comment by nieldeckx** [comment]
@philipdesmedt oracle prices now have 6 decimals. Not that many changes in code (mostly comments and tests changed).

Is it correct that in the following function the collateral-amount set does not really matter but just needs to be larger than the debt?
Screenshot 2021-09-28 at 14 29 23

**Comment by philipdesmedt** [comment]
@nieldeckx that is a good catch lol.. the debt auction is done to sell off DIKO tokens. We need to fetch the price of DIKO from the oracle (or straight from our swap) there since now it is hard-coded to 10 cents.

File: clarity/contracts/arkadiko-auction-engine-v1-1.clar
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-auction-engine-v1-1.clar:254]: Q: we divide by one million here (6 decimals)

what if we handle bitcoin, which has 100M denomination (1 sat = 8 decimals)?

File: clarity/contracts/arkadiko-sip10-reserve-v1-1.clar
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-sip10-reserve-v1-1.clar:41]: Q: where does the division by 10K come from?

**Review by philipdesmedt — CHANGES_REQUESTED** [review]

File: clarity/contracts/arkadiko-auction-engine-v1-1.clar
Comment (nieldeckx) [reply clarity/contracts/arkadiko-auction-engine-v1-1.clar:254]: The oracle now has 6 decimals to save a price. So BTC price in $ will have 6 decimals, and does not depend on the amount of decimals of the token itself.

Comment (philipdesmedt) [reply clarity/contracts/arkadiko-auction-engine-v1-1.clar:254]: mm ok.. so which price do we push to the oracle if bitcoin is $50,000 dollar?

Comment (nieldeckx) [reply clarity/contracts/arkadiko-auction-engine-v1-1.clar:254]: In that case we set $50.000 with 6 decimals so 50000000000.

( Previously you had the price in cents, so 2 decimals )

Comment (philipdesmedt) [reply clarity/contracts/arkadiko-auction-engine-v1-1.clar:254]: OK got it! that makes sense

**Comment by philipdesmedt** [comment]
@nieldeckx let me know when this is ready for review and I'll go through thoroughly (also merge conflict needs fixing)

**Comment by nieldeckx** [comment]
@philipdesmedt  Should be good to review.

- When redeeming a bad debt lot, we will only mint just enough tokens for that lot. 

- The collateral-amount for a bad debt auction does not really matter because when bidding `get-minimum-collateral-amount` will make sure the collateral amount is correctly calculated. So we just need to make sure that the collateral amount on the bad debt auction is definitely enough which is why I set a hard-coded price of $1M per DIKO.

- Both the bidding and redeem function can close a vault. But when closing it could be that there is no debt to burn left which will cause the call to fail. Added a check to avoid burning 0 tokens. But wondering why we want to close a vault in the redeem function? When removing the close-vault call in the redeem function all tests still work. Any idea?

The 3 changes above should be covered in the test.

**Comment by philipdesmedt** [comment]
@nieldeckx great, thanks, will review!

On your last comment, I think I added a call to `close-vault` in the redeem function to optimise things. When someone redeems a lot and we see that the auction can be closed, we do that, so other people or we don't have to take care of this anymore. It's saving an extra on-chain TX. I could be wrong here and maybe it results in bugs tho...

**Comment by nieldeckx** [comment]
@philipdesmedt yeah suggest we keep the `close-vault` call. The extra check prevents it from failing if amount is 0 so all good.

**Review by philipdesmedt — APPROVED** [review]
No comments. Great work!

**Comment by philipdesmedt** [comment]
Don't merge this yet though, I need to test the front-end. Don't think it works yet.

**Comment by philipdesmedt** [comment]
I think this works now. Will merge later

**Comment by philipdesmedt** [comment]
Darn - fixed a few more bugs, both in the oracle changes for the front-end as well as some auction bugs :)

Need to go over `master` again soon once everything is merged.

**Pull request merged**
